module/os/linux/zfs/zpl_file.c

```diff
@@ -743,21 +751,54 @@ zpl_fallocate_common(struct inode *ip, int mode, loff_t offset, loff_t len)
 	spl_inode_lock(ip);
 	olen = i_size_read(ip);
 
-	if (offset > olen) {
-		spl_inode_unlock(ip);
-		return (0);
-	}
-	if (offset + len > olen)
-		len = olen - offset;
-	bf.l_type = F_WRLCK;
-	bf.l_whence = SEEK_SET;
-	bf.l_start = offset;
-	bf.l_len = len;
-	bf.l_pid = 0;
-
 	crhold(cr);
 	cookie = spl_fstrans_mark();
-	error = -zfs_space(ITOZ(ip), F_FREESP, &bf, O_RDWR, offset, cr);
+	if (mode & FALLOC_FL_PUNCH_HOLE) {
+		flock64_t bf;
+
+		if (offset > olen)
+			goto out_unmark;
+
+		if (offset + len > olen)
+			len = olen - offset;
+		bf.l_type = F_WRLCK;
+		bf.l_whence = SEEK_SET;
+		bf.l_start = offset;
+		bf.l_len = len;
+		bf.l_pid = 0;
+
+		error = -zfs_space(ITOZ(ip), F_FREESP, &bf, O_RDWR, offset, cr);
+	} else if ((mode & ~FALLOC_FL_KEEP_SIZE) == 0) {
+		unsigned int percent = zfs_fallocate_reserve_percent;
+		struct kstatfs statfs;
+
+		/* Legacy mode, disable fallocate compatibility. */
+		if (percent == 0) {
+			error = -EOPNOTSUPP;
+			goto out_unmark;
+		}
+
+		/*
+		 * Use zfs_statvfs() instead of dmu_objset_space() since it
+		 * also checks project quota limits, which are relevant here.
+		 */
+		error = zfs_statvfs(ip, &statfs);
```

Comment:
I was surprised to find that zfs_statvfs() gets stats related to a specific file (and its project), as opposed to the "vfs" (filesystem).  As it stands, fallocate() / zfs_statvfs() ignores user/group quotas.  Perhaps as a subject of a future PR, we could rename zfs_statvfs to be more accurate, and also have it take into account user/group quotas.

Reply:
This emulates the behavior of XFS/ext4/others w.r.t. statfs on directories with projectID set on them.  The intent is to set a recursive project ID on a directory tree and treat it like a directory quota/dataset, which is useful for filesystems that don't actually support datasets with their own quota like ZFS.  Changing this for UID/GID quotas might be confusing/surprising for users, maybe not, but would definitely be a big change from all previous history, while the semantics of statfs on directories project quotas has always been this way.

Reply:
Thanks for the context.  Seems like it would be semantically cleaner to just do the more limited change of making fallocate() take into account user/group quotas.